### PR TITLE
NEW : Include sub warehouse in inventory

### PR DESCRIPTION
**Pull request by atm-adrien**
### NEW : Include sub warehouse in inventory

Added the possibility to include sub-warehouses for the calculation of inventories : 

- Creation of a checkbox to choose if the user want to include sub-warehouse for the calculation of an inventory
- When validated an inventory with the previous checkbox checked, sub-warehouse are taken into the calcul of the parent inventory
- Recurring behaviour (warks for sub-sub-warehouse...)
- Conf : INVENTORY_INCLUDE_SUB_WAREHOUSE

image

### CODE REVIEW

File: htdocs/product/inventory/class/inventory.class.php [hunk 1]
```diff
@@ -289,8 +290,15 @@ public function validate(User $user, $notrigger = false)
 				$sql .= " AND ps.fk_product = ".((int) $this->fk_product);
 			}
 			if ($this->fk_warehouse > 0) {
-				$sql .= " AND ps.fk_entrepot = ".((int) $this->fk_warehouse);
+				$sql .= " AND (ps.fk_entrepot = ".((int) $this->fk_warehouse);
 			}
+			if ($this->include_sub_warehouse>0 && $conf->global->INVENTORY_INCLUDE_SUB_WAREHOUSE) {
+				$this->getchildWarehouse($this->fk_warehouse, $TChildWarehouses);
+				foreach ($TChildWarehouses as $childWarehouse) {
+					$sql .= " OR ps.fk_entrepot = ".((int) $childWarehouse);
```
Comment (atm-maxime) [inline htdocs/product/inventory/class/inventory.class.php:298]: Use IN and implode, seems better to me

File: htdocs/install/mysql/migration/15.0.0-16.0.0.sql [hunk 2]
```diff
@@ -237,6 +237,8 @@ ALTER TABLE llx_advtargetemailing RENAME TO llx_mailing_advtarget;
 
 ALTER TABLE llx_mailing ADD UNIQUE uk_mailing(titre, entity);
 
+ALTER TABLE llx_inventory ADD COLUMN include_sub_warehouse boolean DEFAULT FALSE;
```
Comment (atm-maxime) [inline htdocs/install/mysql/migration/15.0.0-16.0.0.sql:240]: Is it necessary to store the value ? Isn't it used only to fill the inventory?

File: htdocs/product/inventory/class/inventory.class.php (hunk 1 above)
Comment (atm-adrien) [reply htdocs/product/inventory/class/inventory.class.php:298]: Done

File: htdocs/install/mysql/migration/15.0.0-16.0.0.sql (hunk 2 above)
Comment (atm-adrien) [reply htdocs/install/mysql/migration/15.0.0-16.0.0.sql]: Done 

image

**Comment by atm-maxime** [comment]
Getpost must not be used in the class, you need to add a new parameter to the validate function

File: htdocs/product/inventory/class/inventory.class.php [hunk 3]
```diff
@@ -264,6 +264,7 @@ public function validate(User $user, $notrigger = false)
 
 		$result = 0;
 
+		$this->include_sub_warehouse = GETPOST('include_sub_warehouse');
```
[the remark this answers is not in the document]
Comment (atm-adrien) [reply htdocs/product/inventory/class/inventory.class.php]: Done

File: htdocs/core/actions_addupdatedelete.inc.php [hunk 4]
```diff
@@ -366,7 +366,14 @@
 
 // Action validate object
 if ($action == 'confirm_validate' && $confirm == 'yes' && $permissiontoadd) {
-	$result = $object->validate($user);
+	if($object->element == 'inventory'){
+		//Case of the conf INVENTORY_INCLUDE_SUB_WAREHOUSE
+		$include_sub_warehouse = !empty(GETPOST('include_sub_warehouse')) ? GETPOST('include_sub_warehouse') : 0;
```
Comment (atm-maxime) [inline htdocs/core/actions_addupdatedelete.inc.php:371]: Not sure this is the best way, let's talk about it

Comment (eldy) [reply htdocs/core/actions_addupdatedelete.inc.php:371]: Move the 
$include_sub_warehouse = !empty(GETPOST('include_sub_warehouse')) ? GETPOST('include_sub_warehouse') : 0;
into the main page and here jsut replace with

`if ($object->element == 'inventory' && !empty($include_sub_warehouse)) {`

Comment (atm-adrien) [reply htdocs/core/actions_addupdatedelete.inc.php]: Done

**Comment by atm-maxime** [comment]
@atm-adrien merge conflict please :)

**Pull request merged**